# NanoComp


Compare multiple runs of Oxford Nanopore sequencing data and alignments

## INSTALLATION
`pip install NanoComp`

## USAGE
```
NanoComp [-h] [-v] [-t THREADS] [--readtype {1D,2D,1D2}] [-o OUTDIR]
                [-p PREFIX]
                [-f {eps,jpeg,jpg,pdf,pgf,png,ps,raw,rgba,svg,svgz,tif,tiff}]
                [-n [NAMES [NAMES ...]]]
                (--fastq [FASTQ [FASTQ ...]] | --summary [SUMMARY [SUMMARY ...]] | --bam [BAM [BAM ...]])


optional arguments:
  -h, --help            show this help message and exit
  -v, --version         Print version and exit.
  -t, --threads THREADS
                        Set the allowed number of threads to be used by the
                        script
  --readtype {1D,2D,1D2}
                        Which read type to extract information about from
                        summary. Options are 1D, 2D, 1D2
  -o, --outdir OUTDIR
                        Specify directory in which output has to be created.
  -p, --prefix PREFIX
                        Specify an optional prefix to be used for the output
                        files.
  -f, --format {eps,jpeg,jpg,pdf,pgf,png,ps,raw,rgba,svg,svgz,tif,tiff}
                        Specify the output format of the plots.
  -n [NAMES [NAMES ...]], --names [NAMES [NAMES ...]]
                        Specify the names to be used for the datasets
  --fastq [FASTQ [FASTQ ...]]
                        Data is in default fastq format.
  --summary [SUMMARY [SUMMARY ...]]
                        Data is a summary file generated by albacore.
  --bam [BAM [BAM ...]]
                        Data as a sorted bam file.
```
